scarb/src/ops/package.rs

```diff
@@ -285,6 +291,39 @@ fn prepare_archive_recipe(pkg: &Package, opts: &PackageOpts) -> Result<ArchiveRe
     Ok(recipe)
 }
 
+fn run_verify(ws: &Workspace<'_>, pkg: &Package, tar: &FileLockGuard) -> Result<()> {
+    ws.config()
+        .ui()
+        .print(Status::new("Verifying", &pkg.id.tarball_name()));
+
+    let decoder: zstd::Decoder<'_, _> = zstd::stream::Decoder::new(tar.deref())?;
+    let mut ar = tar::Archive::new(decoder);
+    let dst = tar.path().parent().unwrap().join(pkg.id.tarball_basename());
+
+    if dst.exists() {
+        remove_dir_all(&dst)?;
+    }
+
+    // From Cargo:
+    // We don't need to set the Modified Time, as it's not relevant to verification
+    // and it errors on filesystems that don't support setting a modified timestamp
+    ar.set_preserve_mtime(false);
+    ar.unpack(dst.parent().unwrap())?;
```

Comment:
What do you think about the following? I don't feel super attached to this as I see quite a bit of changes would be needed to achieve this:

---

I'd rather reuse our existing logic for extraction, that also happens to perform some additional validation: https://github.com/software-mansion/scarb/blob/858300171f3904f2525c2ba41b3d426b61ab8764/scarb/src/core/registry/package_source_store.rs#L43

Foremost: it deals with fs locking, so no concurrent `scarb package` runs will collide.

A nice benefit is that it calls `fs.recreate()` under the hood, so that's less of stuff to reason about here. (+ I wanted to comment on this anyway in the existing code). But on the other hand, we would have to somehow skip the _don't unpack if already did_ logic.

I am thinking of adding such function to `PackageSourceStore`:

```rust
impl PackageSourceStore<'_> {
    #[tracing::instrument(level = "debug", skip_all)]
    pub async fn extract_to(archive: FileLockGuard, fs: Filesystem, config: &Config) -> Result<Utf8PathBuf>
}
```